src/Microsoft.ML.Data/Transforms/OneToOneTransformerBase.cs

```diff
@@ -0,0 +1,177 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.ML.Core.Data;
+using Microsoft.ML.Runtime.Model;
+
+namespace Microsoft.ML.Runtime.Data
+{
+    public abstract class OneToOneTransformerBase : ITransformer, ICanSaveModel
+    {
+        protected readonly IHost Host;
+        protected readonly (string input, string output)[] ColumnPairs;
+
+        protected OneToOneTransformerBase(IHost host, (string input, string output)[] columns)
+        {
+            Contracts.AssertValue(host);
+            host.CheckValue(columns, nameof(columns));
+
+            var newNames = new HashSet<string>();
+            foreach (var column in columns)
+            {
+                host.CheckNonEmpty(column.input, nameof(columns));
+                host.CheckNonEmpty(column.output, nameof(columns));
+
+                if (!newNames.Add(column.output))
+                    throw Contracts.ExceptParam(nameof(columns), $"Output column '{column.output}' specified multiple times");
+            }
+
+            Host = host;
+            ColumnPairs = columns;
+        }
+
+        protected OneToOneTransformerBase(IHost host, ModelLoadContext ctx)
+        {
+            Host = host;
+            // *** Binary format ***
+            // int: number of added columns
+            // for each added column
+            //   int: id of output column name
+            //   int: id of input column name
+
+            int n = ctx.Reader.ReadInt32();
+            ColumnPairs = new (string input, string output)[n];
+            for (int i = 0; i < n; i++)
+            {
+                string output = ctx.LoadNonEmptyString();
+                string input = ctx.LoadNonEmptyString();
+                ColumnPairs[i] = (input, output);
+            }
+        }
+
+        public abstract void Save(ModelSaveContext ctx);
+
+        protected void SaveColumns(ModelSaveContext ctx)
+        {
+            Host.CheckValue(ctx, nameof(ctx));
+
+            // *** Binary format ***
+            // int: number of added columns
+            // for each added column
+            //   int: id of output column name
+            //   int: id of input column name
+
+            ctx.Writer.Write(ColumnPairs.Length);
+            for (int i = 0; i < ColumnPairs.Length; i++)
+            {
+                ctx.SaveNonEmptyString(ColumnPairs[i].output);
+                ctx.SaveNonEmptyString(ColumnPairs[i].input);
+            }
+        }
+
+        private void CheckInput(ISchema inputSchema, int col, out int srcCol)
```

Comment:
>private [](start = 8, length = 7)

make this public and call it in GetOutputSchema in each estimator.

Reply:
Cannot do this, because GetOutputSchema in the estimator operates over `SchemaShape`

---
In reply to: [213863333](https://github.com/dotnet/machinelearning/pull/753#discussion_r213863333) [](ancestors = 213863333)

Reply:
if only ISchema and SchemaShape were relatives....

---
In reply to: [213863552](https://github.com/dotnet/machinelearning/pull/753#discussion_r213863552) [](ancestors = 213863552,213863333)